rust/datafusion/src/physical_plan/functions.rs

```diff
@@ -702,14 +1197,912 @@ mod tests {
+            ],
+            Some("phabet"),
+        )?;
+        generic_string_function(
+            BuiltinScalarFunction::Substr,
+            vec![
+                lit(ScalarValue::Utf8(Some("alphabet".to_string()))),
+                lit(ScalarValue::Int64(None)),
+                lit(ScalarValue::Int64(Some(20))),
+            ],
+            None,
+        )?;
+        generic_string_function(
+            BuiltinScalarFunction::Substr,
+            vec![
+                lit(ScalarValue::Utf8(Some("alphabet".to_string()))),
+                lit(ScalarValue::Int64(Some(3))),
+                lit(ScalarValue::Int64(None)),
+            ],
+            None,
+        )?;
+        generic_string_function(
+            BuiltinScalarFunction::Translate,
+            vec![
+                lit(ScalarValue::Utf8(Some("12345".to_string()))),
+                lit(ScalarValue::Utf8(Some("143".to_string()))),
+                lit(ScalarValue::Utf8(Some("ax".to_string()))),
+            ],
+            Some("a2x5"),
+        )?;
+        generic_string_function(
+            BuiltinScalarFunction::Translate,
+            vec![
+                lit(ScalarValue::Utf8(None)),
+                lit(ScalarValue::Utf8(Some("143".to_string()))),
+                lit(ScalarValue::Utf8(Some("ax".to_string()))),
+            ],
+            None,
+        )?;
+        generic_string_function(
+            BuiltinScalarFunction::Translate,
+            vec![
+                lit(ScalarValue::Utf8(Some("12345".to_string()))),
+                lit(ScalarValue::Utf8(None)),
+                lit(ScalarValue::Utf8(Some("ax".to_string()))),
+            ],
+            None,
+        )?;
+        generic_string_function(
+            BuiltinScalarFunction::Translate,
+            vec![
+                lit(ScalarValue::Utf8(Some("12345".to_string()))),
+                lit(ScalarValue::Utf8(Some("143".to_string()))),
+                lit(ScalarValue::Utf8(None)),
+            ],
+            None,
+        )?;
+        generic_string_function(
+            BuiltinScalarFunction::Upper,
+            vec![lit(ScalarValue::Utf8(Some("upper".to_string())))],
+            Some("UPPER"),
+        )?;
+        generic_string_function(
+            BuiltinScalarFunction::Upper,
+            vec![lit(ScalarValue::Utf8(Some("UPPER".to_string())))],
+            Some("UPPER"),
+        )?;
+        generic_string_function(
+            BuiltinScalarFunction::Upper,
+            vec![lit(ScalarValue::Utf8(None))],
+            None,
+        )?;
+        generic_string_function(
+            BuiltinScalarFunction::ToHex,
+            vec![lit(ScalarValue::Int32(Some(2147483647)))],
+            Some("7fffffff"),
+        )?;
+        generic_string_function(
+            BuiltinScalarFunction::ToHex,
+            vec![lit(ScalarValue::Int32(None))],
+            None,
+        )?;
+
+        Ok(())
+    }
+
+    fn generic_string_i32_function(
```

Comment:
I have done a major update since you reviewed so all functions in `functions.rs` now use a single macro which also checks expected errors. I had planned to do this but never got to it.